### PR TITLE
[input] Re-factor remote control

### PR DESCRIPTION
**Pull request by afedchin**
in order of preparation to an implementation of support remote control on xbox I've made refactoring current code.

### CODE REVIEW

**Comment by afedchin** [comment]
jenkins build this please

**Comment by afedchin** [comment]
jenkins build this please

**Comment by afedchin** [comment]
jenkins build this please

**Comment by afedchin** [comment]
jenkins build this please

**Comment by afedchin** [comment]
jenkins build this please

File: xbmc/input/InputManager.h [hunk 1]
```diff
@@ -208,6 +203,12 @@ class CInputManager : public ISettingCallback,
    */
   void InitializeRemoteControl();
 
+  /*! \brief Check if the remote control exists
+   *
+   * \return true if remote control is enabled, false otherwise 
+   */
+  bool IsRemoteControlExists();
```
Comment (garbear) [inline xbmc/input/InputManager.h]:
```c++
bool HasRemoteControl();
```

File: xbmc/input/IRTranslator.h [hunk 2]
```diff
@@ -34,7 +34,7 @@ class CIRTranslator
   /*!
    * \brief Loads Lircmap.xml/IRSSmap.xml
    */
-  void Load();
+  void Load(std::string irMapName);
```
Comment (garbear) [inline xbmc/input/IRTranslator.h]:
```c++
void Load(const std::string &irMapName);
```

File: xbmc/input/remote/IRemoteControl.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+   */
+  virtual void Initialize() = 0;
+
+  /*!
+   * \brief Deinitialize remote control
+   */
+  virtual void Disconnect() = 0;
+
+  /*!
+   * \brief Get the button readed from IR
```
Comment (garbear) [inline xbmc/input/remote/IRemoteControl.h]: `Get the button read from IR`

File: xbmc/input/remote/IRemoteControl.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+   *
+   * \return A button code
+   */
+  virtual unsigned short GetButton() const = 0;
```
Comment (garbear) [inline xbmc/input/remote/IRemoteControl.h]: If bit width is relevant, can you use `uint16_t` and include <stdint.h>?

File: xbmc/input/remote/IRemoteControl.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+  /*!
+   * \brief Get the time while a button was hold
+   *
+   * \return A hold time
```
Comment (garbear) [inline xbmc/input/remote/IRemoteControl.h]: Plus 1 point for documenting the units, plus 2 points for including units in function name

e.g.
```c++
virtual unsigned int GetHoldTimeMs() const = 0;
```

File: xbmc/input/remote/IRemoteControl.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  /*!
+   * \brief Get the initialized status
+   */
+  virtual bool IsInitialized() const = 0;
+
+  /*!
+   * \brief Set enabled status
+   */
+  virtual void SetEnabled(bool) { }
```
Comment (garbear) [inline xbmc/input/remote/IRemoteControl.h]: Can you include a parameter name? That way, when this function is copied to multiple interfaces, we can remain consistent and not end up with a bunch of different names

e.g.
```c++
virtual void SetEnabled(bool bEnabled) { }
```

Same below

File: xbmc/input/remote/IRemoteControl.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+  /*!
+   * \brief Add a command to send to a bus
+   */
+  virtual void AddSendCommand(const std::string&) { }
+
+  /*!
+   * \brief Get the in-use status
+   */
+  virtual bool IsInUse() const { return false; }
```
Comment (garbear) [inline xbmc/input/remote/IRemoteControl.h]: Why did you specify a definition and make these four commands optional? I'm not against it, just wondering why. I tend to prefer my interfaces pure-virtual, it result in more code but makes implementers conscious of their decision to omit definitions.

File: xbmc/platform/linux/input/LIRC.h [hunk 8]
```diff
@@ -24,28 +24,33 @@
...
-  bool IsInUse() const { return m_used; }
-  bool IsInitialized() const { return m_bInitialized; }
-  void AddSendCommand(const std::string& command);
+  unsigned int GetHoldTime() const override;
+  void SetDeviceName(const std::string& value) override;
+  void SetEnabled(bool value) override;
+  bool IsInUse() const override { return m_used; }
+  bool IsInitialized() const override { return m_bInitialized; }
+  void AddSendCommand(const std::string& command) override;
+  std::string GetMapFile() override { return "Lircmap.xml"; }
```
Comment (garbear) [inline xbmc/platform/linux/input/LIRC.h]: Can you move this definition to the .cpp file and make `"Lircmap.xml"` a #define?

File: xbmc/input/remote/IRemoteControl.h [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+ *  along with this Program; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+
+/*!
+  * \brief Interface for remote control
+  */
+class IRemoteControl
```
Comment (garbear) [inline xbmc/input/remote/IRemoteControl.h]: Can you put this in namespace `KODI::REMOTE`?

**Comment by garbear** [comment]
Really great job! All my comments are 1-liners

File: xbmc/input/remote/IRemoteControl.h [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+  /*!
+   * \brief Deinitialize remote control
+   */
+  virtual void Disconnect() = 0;
+
+  /*!
+   * \brief Get the button readed from IR
+   *
+   * \return A button code
```
Comment (garbear) [inline xbmc/input/remote/IRemoteControl.h]: Can you clarify that this code comes from XBIRRemote.h?

**Comment by garbear** [comment]
A single request to make this refactoring even more stellar: Can you move XBIRRemote.h to your new remotes/ subdir? Maybe rename it to not include XB? thanks!

File: xbmc/platform/linux/input/LIRC.cpp [hunk 11]
```diff
@@ -61,7 +73,12 @@ CRemoteControl::~CRemoteControl()
     fclose(m_file);
 }
 
-void CRemoteControl::SetEnabled(bool value)
+std::string CRemoteControl::GetMapFile()
+{
+  return LIRC_MAP_FILENAME;
+}
+
+void CRemoteControl::SetEnabled(bool bEnabled)
```
Comment (Rechi) [inline xbmc/platform/linux/input/LIRC.cpp]: `bool value` or alue has to be replaced with `bEnabled` in the whole function

Comment (afedchin) [reply xbmc/platform/linux/input/LIRC.cpp]: oops. thanks.

File: xbmc/input/remote/IRemoteControl.h [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+   */
+  virtual void AddSendCommand(const std::string&) = 0;
+
+  /*!
+   * \brief Get the in-use status
+   */
+  virtual bool IsInUse() const = 0;
+};
+  }
+}
```
Comment (garbear) [inline xbmc/input/remote/IRemoteControl.h]: Missing newline (this causes headaches for people with IDEs that automatically enforce this)

File: xbmc/input/remote/IRemoteControl.h [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+  /*!
+   * \brief Set a name for a device
+   */
+  virtual void SetDeviceName(const std::string&) = 0;
+
+  /*!
+   * \brief Add a command to send to a bus
+   */
+  virtual void AddSendCommand(const std::string&) = 0;
```
Comment (garbear) [inline xbmc/input/remote/IRemoteControl.h]: Parameter name is still missing

File: xbmc/input/remote/IRemoteControl.h (hunk 12 above)
Comment (afedchin) [reply xbmc/input/remote/IRemoteControl.h]: fixed. thanks

**Review by garbear — APPROVED** [review]

**Comment by afedchin** [comment]
if there is no more objection this will go tonight

**Comment by garbear** [comment]
go now

**Comment by MilhouseVH** [comment]
This PR is causing a problem with remote input after resuming from suspend (on Linux, specifically LibreELEC).

Starting with build [#0210](https://forum.kodi.tv/showthread.php?tid=298462&pid=2702465#pid2702465), when device (eg. x86_64 Revo 3700 with USB MCE IR) is resumed from suspend Kodi no longer receives any remote control inputs.

I've built the latest master Kodi with and without this PR, and the version with this PR has the resume-from-suspend issue while the version without works perfectly after resuming.

Log without PR: http://ix.io/Gjw (works perfectly after resume)
Log with PR: http://ix.io/Gjw (no remote input seen after resume)

What seems to be missing in the "bad" log is the restart of lirc - this is what happens in the "good" log after resuming:
```
08:18:28.901 T:139900420427136  NOTICE: OnWake: Running resume jobs
08:18:28.901 T:139900420427136   DEBUG: ------ Window Deinit (DialogBusy.xml) ------
08:18:28.901 T:139900420427136  NOTICE: OnWake: Restarting lirc
08:18:28.903 T:139898102593280   DEBUG: Thread RemoteControl start, auto delete: false
08:18:28.903 T:139898102593280    INFO: LIRC Process: using: /run/lirc/lircd
08:18:28.904 T:139898102593280    INFO: LIRC Connect: successfully started
08:18:28.905 T:139898102593280   DEBUG: Thread RemoteControl 139898102593280 terminating
```

**Comment by afedchin** [comment]
> Log without PR: http://ix.io/Gjw (works perfectly after resume)
> Log with PR: http://ix.io/Gjw (no remote input seen after resume)

the links are identical

**Comment by MilhouseVH** [comment]
Sorry, fat fingers:

Log without PR: http://ix.io/Gjw (works perfectly after resume)
Log with PR: http://ix.io/GjB (no remote input seen after resume)

**Comment by afedchin** [comment]
Should work now, see https://github.com/xbmc/xbmc/commit/5c9c31804a20c37b346e59b52f1e2dda26789b3c

**Comment by MilhouseVH** [comment]
Thanks, https://github.com/xbmc/xbmc/commit/5c9c31804a20c37b346e59b52f1e2dda26789b3c fixes the issue.

**Pull request merged**